### PR TITLE
CDAP-3902: Remove remote call for registering dataset usage in DatasetInstanceService

### PR DESCRIPTION
**Pull request by jwang47**
- Added DatasetProvider and DatasetMetaProvider
- Made UsageRegistry obtain the usage registry dataset instance by using
  DatasetProvider instead of DatasetFramework

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetMetaProvider.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.data2.datafabric.dataset;
+
+import co.cask.cdap.api.dataset.DatasetSpecification;
+import co.cask.cdap.proto.DatasetMeta;
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetMetaProvider.java]: Unused import

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetProvider.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        DatasetModule module = DatasetModules.getDatasetModule(moduleClass);
+        module.register(registry);
+      } catch (Exception e) {
+        LOG.error("Was not able to load dataset module class {} while trying to load type {}",
+                  moduleMeta.getClassName(), implementationInfo, e);
+        throw Throwables.propagate(e);
+      }
+    }
+
+    // contract of DatasetTypeMeta is that the last module returned by getModules() is the one
```
Comment (rohitsinha54) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetProvider.java]: Is this true for custom dataset which embed other custom dataset ? Will the dataset type will appear after the embedded one ? or the embedded one  afterwards ?

Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetProvider.java]: Not sure, I just moved this code from somewhere else.

File: cdap-master/src/main/java/SimpleTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ *
+ */
+public class SimpleTest {
```
Comment (poornachandra) [inline cdap-master/src/main/java/SimpleTest.java]: This class is needs to be removed.

**Comment by jwang47** [comment]
Reopened https://github.com/caskdata/cdap/pull/4320

**Pull request closed without merging**